tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,512 @@
+#include <utils/snapmgr.h>
+#include <nodes/memnodes.h>
+#include <storage/lockdefs.h>
+#include <access/htup_details.h>
+#include <access/htup.h>
+#include <access/xact.h>
+
+#include <catalog.h>
+#include <scanner.h>
+#include <scan_iterator.h>
+
+#include "continuous_agg.h"
+#include "continuous_aggs/materialize.h"
+#include "invalidation.h"
+
+typedef struct CaggInvalidationState
+{
+	ContinuousAgg cagg;
+	MemoryContext per_tuple_mctx;
+	Relation cagg_log_rel;
+	Snapshot snapshot;
+} CaggInvalidationState;
+
+typedef enum LogType
+{
+	LOG_HYPER,
+	LOG_CAGG,
+} LogType;
+
+static Relation
+open_invalidation_log(LogType type, LOCKMODE lockmode)
+{
+	static const CatalogTable logmappings[] = {
+		[LOG_HYPER] = CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG,
+		[LOG_CAGG] = CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+	};
+	Catalog *catalog = ts_catalog_get();
+	Oid relid = catalog_get_table_id(catalog, logmappings[type]);
+
+	return table_open(relid, lockmode);
+}
+
+static void
+cagg_scan_by_hypertable_init(ScanIterator *iterator, int32 hyper_id, LOCKMODE lockmode)
+{
+	*iterator = ts_scan_iterator_create(CONTINUOUS_AGG, lockmode, CurrentMemoryContext);
+	iterator->ctx.index =
+		catalog_get_index(ts_catalog_get(), CONTINUOUS_AGG, CONTINUOUS_AGG_RAW_HYPERTABLE_ID_IDX);
+	ts_scan_iterator_scan_key_init(iterator,
+								   Anum_continuous_agg_raw_hypertable_id_idx_raw_hypertable_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(hyper_id));
+}
+
+/*
+ * Get a list of continuous aggregate IDs for a hypertable.
+ *
+ * Since this is just an integer list, the memory cost is not big even if
+ * there are a lot of continuous aggregates.
+ */
+static List *
+get_cagg_ids(int32 hyper_id)
+{
+	List *cagg_ids = NIL;
+	ScanIterator iterator;
+
+	cagg_scan_by_hypertable_init(&iterator, hyper_id, AccessShareLock);
+
+	ts_scanner_foreach(&iterator)
+	{
+		bool isnull;
+		TupleInfo *ti_cagg = ts_scan_iterator_tuple_info(&iterator);
+		Datum cagg_hyper_id =
+			slot_getattr(ti_cagg->slot, Anum_continuous_agg_mat_hypertable_id, &isnull);
+		Assert(!isnull);
+		cagg_ids = lappend_int(cagg_ids, DatumGetInt32(cagg_hyper_id));
+	}
+
+	ts_scan_iterator_close(&iterator);
+
+	return cagg_ids;
+}
+
+static void
+hypertable_invalidation_scan_init(ScanIterator *iterator, int32 hyper_id, LOCKMODE lockmode)
+{
+	*iterator = ts_scan_iterator_create(CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG,
+										lockmode,
+										CurrentMemoryContext);
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(),
+											CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG,
+											CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG_IDX);
+	ts_scan_iterator_scan_key_init(
+		iterator,
+		Anum_continuous_aggs_hypertable_invalidation_log_idx_hypertable_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(hyper_id));
+}
+
+static HeapTuple
+create_invalidation_tup(const TupleDesc tupdesc, int32 cagg_hyper_id, int64 modtime, int64 start,
```

Comment:
In this code, it is consistently used for pointers that should not be modified. I don't use const for pass by reference since it doesn't serve much purpose.